File: CS/WinForms.Client/EditForm.cs

using DevExpress.XtraEditors;

namespace WinForms.Client {
    public partial class EditForm : XtraForm {
        public EditForm() {
            InitializeComponent();
        }
        BindingSource bsDataItem = new BindingSource();
        public void SetEditableData(OData.Server.BusinessObjects.OrderItem item) {
            bsDataItem.DataSource = DTO.Create(item);
            dataLayoutControl1.DataSource = bsDataItem;
        }
        public void ApplyChanges(OData.Server.BusinessObjects.OrderItem item) {
            var dto = (DTO)bsDataItem.DataSource;
            dto.ApplyChanges(item);
        }
        void Cancel_button_Click(object sender, EventArgs e) {
            this.DialogResult = DialogResult.Cancel;
        }
        void OK_button_Click(object sender, EventArgs e) {
            this.DialogResult = DialogResult.OK;
        }
    }
}

File: VB/WinForms.Client/EditForm.vb
Imports DevExpress.XtraEditors

Namespace WinForms.Client

    Public Partial Class EditForm
        Inherits XtraForm

        Public Sub New()
            InitializeComponent()
        End Sub

        Private bsDataItem As BindingSource = New BindingSource()

        Public Sub SetEditableData(ByVal item As OData.Server.BusinessObjects.OrderItem)
            bsDataItem.DataSource = DTO.Create(item)
            dataLayoutControl1.DataSource = bsDataItem
        End Sub

        Public Sub ApplyChanges(ByVal item As OData.Server.BusinessObjects.OrderItem)
            Dim dto = CType(bsDataItem.DataSource, DTO)
            dto.ApplyChanges(item)
        End Sub

        Private Sub Cancel_button_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.DialogResult = DialogResult.Cancel
        End Sub

        Private Sub OK_button_Click(ByVal sender As Object, ByVal e As EventArgs)
            Me.DialogResult = DialogResult.OK
        End Sub
    End Class
End Namespace
